graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -107,26 +117,148 @@ def rename_schema(
+
+
+def _check_for_cascading_type_suppression(
+    schema_ast: DocumentNode, renamings: Mapping[str, Optional[str]], query_type: str
+) -> None:
+    """Check for fields with suppressed types or unions whose members were all suppressed."""
+    visitor = CascadingSuppressionCheckVisitor(renamings, query_type)
+    visit(schema_ast, visitor)
+    if visitor.fields_to_suppress or visitor.union_types_to_suppress:
+        error_message_components = [
+            f"Type renamings {renamings} would require further suppressions to produce a valid"
+            f"renamed schema."
+        ]
+        if visitor.fields_to_suppress:
+            for object_type in visitor.fields_to_suppress:
+                error_message_components.append(f"Object type {object_type} contains: ")
+                error_message_components += [
+                    f"field {field} of suppressed type "
+                    f"{visitor.fields_to_suppress[object_type][field]}, "
+                    for field in visitor.fields_to_suppress[object_type]
+                ]
+            error_message_components.append(
+                "A schema containing a field that is of a nonexistent type is invalid. When field "
+                "suppression is supported, you can fix this problem by suppressing the fields "
+                "shown above."
+            )
+        if visitor.union_types_to_suppress:
+            for union_type in visitor.union_types_to_suppress:
+                if union_type.types is None:
+                    union_name = union_type.name.value
+                    # This shouldn't happen, but the GraphQL core library type-hints the field as
+                    # Optional[FrozenList[NamedTypeNode]]. Follow-up discussion here:
+                    # https://github.com/graphql-python/graphql-core/issues/98
+                    raise AssertionError(
+                        f"Union {union_name} must include one or more unique member types, but "
+                        f"{union_name}'s list of union types was None."
+                    )
+                error_message_components.append(
+                    f"Union type {union_type} has no non-suppressed members: "
+                )
+                error_message_components += [
+                    union_member.name.value for union_member in union_type.types
+                ]
+            error_message_components.append(
+                "To fix this, you can suppress the union as well by adding `union_type: None` to "
+                "the `renamings` argument of `rename_schema`, for each value of `union_type` "
+                "described here. Note that adding suppressions may lead to other types, fields, "
+                "unions, etc. requiring suppression so you may need to iterate on this before "
+                "getting a legal schema."
+            )
+        raise CascadingSuppressionError("\n".join(error_message_components))
```

Comment:
Consider, as a quality-of-life improvement (and perhaps in a subsequent PR), adding a way to programmatically read what went wrong that doesn't require parsing the error message text. When defining this exception, you get to specify what arguments its constructor takes, and what attributes it exposes — so you could in principle have a list of appropriately-typed values that describe what went wrong: for example, for each suppression, what additional steps must be taken to make it valid.

This would allow a pleasant UX workflow: attempt a renaming, observe it fail with constructive feedback, select which of the required changes you'd like to apply and for which you'd instead like to abandon the renaming, apply the prerequisite changes, then apply the renaming itself. The computer did almost all the work, and the user essentially only had to do the decision-making.

Reply:
I like that idea! Coming soon :)